File: net/OpenDraft.Lang/MessageUtility.cs
// <copyright file="MessageUtility.cs" company="Jon Rowlett">
// Copyright (c) Jon Rowlett. All rights reserved.
// </copyright>

namespace OpenDraft.Lang;

/// <summary>
/// Helpers for creating messages produced by the language components.
/// </summary>
internal static class MessageUtility
{
    /// <summary>
    /// Represents the message id for an unterminated string literal error message.
    /// </summary>
    public static readonly MessageId UnterminatedStringLiteralMessageId = new MessageId(Constants.DefaultMessageCategory, 1000);

    /// <summary>
    /// Represents the message id for an invalid character error message.
    /// </summary>
    public static readonly MessageId InvalidCharacterMessageId = new MessageId(Constants.DefaultMessageCategory, 1001);

    /// <summary>
    /// Represents the message id for an unterminated multi-line comment error message.
    /// </summary>
    public static readonly MessageId UnterminatedMultiLineCommentMessageId = new MessageId(Constants.DefaultMessageCategory, 1002);

    /// <summary>
    /// Represents the message id for an unknown escape sequence error message.
    /// </summary>
    public static readonly MessageId UnknownEscapeSequenceMessageId = new MessageId(Constants.DefaultMessageCategory, 1003);

    /// <summary>
    /// Represents the message id for an invalid hexadecimal escape sequence error message.
    /// </summary>
    public static readonly MessageId InvalidHexEscapeSequenceMessageId = new MessageId(Constants.DefaultMessageCategory, 1004);

    /// <summary>
    /// Represents the message id for an unexpected end of file error message.
    /// </summary>
    public static readonly MessageId UnexpectedEndOfFileMessageId = new MessageId(Constants.DefaultMessageCategory, 2000);

    /// <summary>
    /// Represents the message id for a keyword expected error message.
    /// </summary>
    public static readonly MessageId KeywordExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2001);

    /// <summary>
    /// Represents the message id for a symbol expected error message.
    /// </summary>
    public static readonly MessageId SymbolExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2002);

    /// <summary>
    /// Represents the message id for an identifier expected error message.
    /// </summary>
    public static readonly MessageId IdentifierExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2003);

    /// <summary>
    /// Represents the message id for a string literal expected error message.
    /// </summary>
    public static readonly MessageId StringLiteralExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2004);

    /// <summary>
    /// Represents the message id for a numeric literal expected error message.
    /// </summary>
    public static readonly MessageId NumericLiteralExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2005);

    /// <summary>
    /// Represents the message id for an integer literal expected error message.
    /// </summary>
    public static readonly MessageId IntegerLiteralExpectedMessageId = new MessageId(Constants.DefaultMessageCategory, 2006);

    /// <summary>
    /// Represents the message id for an unexpected token error message.
    /// </summary>
    public static readonly MessageId UnexpectedTokenMessageId = new MessageId(Constants.DefaultMessageCategory, 2100);

    /// <summary>
    /// Creates a message for an unterminated string literal error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <returns>The created message.</returns>
    public static Message UnterminatedStringLiteral(SourceReference source)
    {
        return new Message(
            MessageSeverity.Error,
            UnterminatedStringLiteralMessageId,
            source,
            "Unterminated string literal.");
    }

    /// <summary>
    /// Creates a message for an invalid character error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <param name="character">The invalid character.</param>
    /// <returns>The created message.</returns>
    public static Message InvalidCharacter(SourceReference source, char character)
    {
        return new Message(
            MessageSeverity.Error,
            InvalidCharacterMessageId,
            source,
            $"Invalid character '{character}'.");
    }

    /// <summary>
    /// Creates a message for an unterminated multi-line comment error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <returns>The created message.</returns>
    public static Message UnterminatedMultiLineComment(SourceReference source)
    {
        return new Message(
            MessageSeverity.Error,
            UnterminatedMultiLineCommentMessageId,
            source,
            "Unterminated multi-line comment.");
    }

    /// <summary>
    /// Creates a message for an unknown escape sequence error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <param name="character">The unknown escape sequence character.</param>
    /// <returns>The created message.</returns>
    public static Message UnknownEscapeSequence(SourceReference source, char character)
    {
        return new Message(
            MessageSeverity.Error,
            UnknownEscapeSequenceMessageId,
            source,
            $"Unknown escape sequence '\\{character}'.");
    }

    /// <summary>
    /// Creates a message for an invalid hexadecimal escape sequence error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <returns>The created message.</returns>
    public static Message InvalidHexEscapeSequence(SourceReference source)
    {
        return new Message(
            MessageSeverity.Error,
            InvalidHexEscapeSequenceMessageId,
            source,
            "Invalid hexadecimal escape sequence.");
    }

    /// <summary>
    /// Creates a message for an unexpected end of file error.
    /// </summary>
    /// <param name="source">The source reference for the error.</param>
    /// <returns>The created message.</returns>
    public static Message UnexpectedEndOfFile(SourceReference source)
    {
        return new Message(
            MessageSeverity.Error,
            UnexpectedEndOfFileMessageId,
            source,
            "Unexpected end of file.");
    }

    /// <summary>
    /// Creates a message for a keyword expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <param name="expected">The expected keyword.</param>
    /// <returns>The created message.</returns>
    public static Message KeywordExpected(
        Token? token,
        SourceReference fileSource,
        Keyword expected)
    {
        return new Message(
            MessageSeverity.Error,
            KeywordExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected keyword '{expected}', but found '{token?.ToString() ?? "end of file"}' which is not a keyword.");
    }

    /// <summary>
    /// Creates a message for a symbol expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <param name="expected">The expected symbol.</param>
    /// <returns>The created message.</returns>
    public static Message SymbolExpected(
        Token? token,
        SourceReference fileSource,
        Symbol expected)
    {
        return new Message(
            MessageSeverity.Error,
            SymbolExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected symbol '{expected}', but found '{token?.ToString() ?? "end of file"}' which is not a symbol.");
    }

    /// <summary>
    /// Creates a message for an identifier expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <returns>The created message.</returns>
    public static Message IdentifierExpected(
        Token? token,
        SourceReference fileSource)
    {
        return new Message(
            MessageSeverity.Error,
            IdentifierExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected identifier, but found '{token?.ToString() ?? "end of file"}' which is not an identifier.");
    }

    /// <summary>
    /// Creates a message for a string literal expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <returns>The created message.</returns>
    public static Message StringLiteralExpected(
        Token? token,
        SourceReference fileSource)
    {
        return new Message(
            MessageSeverity.Error,
            StringLiteralExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected string literal, but found '{token?.ToString() ?? "end of file"}' which is not a string literal.");
    }

    /// <summary>
    /// Creates a message for a numeric literal expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <returns>The created message.</returns>
    public static Message NumericLiteralExpected(
        Token? token,
        SourceReference fileSource)
    {
        return new Message(
            MessageSeverity.Error,
            NumericLiteralExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected numeric literal, but found '{token?.ToString() ?? "end of file"}' which is not a numeric literal.");
    }

    /// <summary>
    /// Creates a message for an integer literal expected error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <returns>The created message.</returns>
    public static Message IntegerLiteralExpected(
        Token? token,
        SourceReference fileSource)
    {
        return new Message(
            MessageSeverity.Error,
            IntegerLiteralExpectedMessageId,
            token?.Source ?? fileSource,
            $"Expected integer literal, but found '{token?.ToString() ?? "end of file"}' which is not an integer literal.");
    }

    /// <summary>
    /// Creates a message for an unexpected token error.
    /// </summary>
    /// <param name="token">The token that caused the error.</param>
    /// <param name="fileSource">The source reference for the file being parsed, used if the token is null.</param>
    /// <returns>The created message.</returns>
    public static Message UnexpectedToken(
        Token? token,
        SourceReference fileSource)
    {
        return new Message(
            MessageSeverity.Error,
            UnexpectedTokenMessageId,
            token?.Source ?? fileSource,
            $"Unexpected token '{token?.ToString() ?? "end of file"}'.");
    }
}

File: cpp/src/lang/message_utility.cpp
/**
 * Copyright (c) Jon Rowlett. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 * @file message_utility.cpp
 * @brief Implements the message_utility class and related functions.
 */

#include "message_utility.hpp"

namespace opendraft::lang
{
    constexpr char DEFAULT_CATEGORY[] = "ODL";
    const message_id message_utility::UNTERMINATED_STRING_LITERAL_ID(DEFAULT_CATEGORY, 1000);
    const message_id message_utility::UNRECOGNIZED_CHARACTER_ID(DEFAULT_CATEGORY, 1001);
    const message_id message_utility::UNTERMINATED_MULTI_LINE_COMMENT_ID(DEFAULT_CATEGORY, 1002);
    const message_id message_utility::UNKNOWN_ESCAPE_SEQUENCE_ID(DEFAULT_CATEGORY, 1003);
    const message_id message_utility::INVALID_HEX_ESCAPE_SEQUENCE_ID(DEFAULT_CATEGORY, 1004);
    const message_id message_utility::EXPECTED_DECIMAL_EXPONENT_ID(DEFAULT_CATEGORY, 1005);
    const message_id message_utility::INVALID_UNIT_SPECIFIER_ID(DEFAULT_CATEGORY, 1006);
    const message_id message_utility::INVALID_NUMBER_AFTER_RADIX_ID(DEFAULT_CATEGORY, 1007);

    const message_id message_utility::UNEXPECTED_END_OF_FILE_ID(DEFAULT_CATEGORY, 2000);
    const message_id message_utility::KEYWORD_EXPECTED_ID(DEFAULT_CATEGORY, 2001);
    const message_id message_utility::SYMBOL_EXPECTED_ID(DEFAULT_CATEGORY, 2002);
    const message_id message_utility::IDENTIFIER_EXPECTED_ID(DEFAULT_CATEGORY, 2003);
    const message_id message_utility::STRING_LITERAL_EXPECTED_ID(DEFAULT_CATEGORY, 2004);
    const message_id message_utility::NUMERIC_LITERAL_EXPECTED_ID(DEFAULT_CATEGORY, 2005);
    const message_id message_utility::INTEGER_LITERAL_EXPECTED_ID(DEFAULT_CATEGORY, 2006);
    const message_id message_utility::VARIABLE_NAME_EXPECTED_AFTER_TYPE_REFERENCE_ID(DEFAULT_CATEGORY, 2007);
    const message_id message_utility::UNEXPECTED_TOKEN_ID(DEFAULT_CATEGORY, 2100);

    message message_utility::unterminated_string_literal(const source_reference& source)
    {
        return message(
            source,
            UNTERMINATED_STRING_LITERAL_ID,
            severity::SEVERITY_ERROR,
            "unterminated string literal");
    }

    message message_utility::unrecognized_character(const source_reference& source, char ch)
    {
        return message(
            source,
            UNRECOGNIZED_CHARACTER_ID,
            severity::SEVERITY_ERROR,
            "unrecognized character");
    }

    message message_utility::unterminated_multi_line_comment(const source_reference& source)
    {
        return message(
            source,
            UNTERMINATED_MULTI_LINE_COMMENT_ID,
            severity::SEVERITY_ERROR,
            "unterminated multi-line comment");
    }

    message message_utility::unknown_escape_sequence(const source_reference& source, char ch)
    {
        return message(
            source,
            UNKNOWN_ESCAPE_SEQUENCE_ID,
            severity::SEVERITY_ERROR,
            "unknown escape sequence");
    }

    message message_utility::invalid_hex_escape_sequence(const source_reference& source, char ch)
    {
        return message(
            source,
            INVALID_HEX_ESCAPE_SEQUENCE_ID,
            severity::SEVERITY_ERROR,
            "invalid hex escape sequence");
    }

    message message_utility::expected_decimal_exponent(const source_reference& source)
    {
        return message(
            source,
            EXPECTED_DECIMAL_EXPONENT_ID,
            severity::SEVERITY_ERROR,
            "expected decimal exponent");
    }

    message message_utility::invalid_unit_specifier(const source_reference& source, const std::string_view& unit)
    {
        return message(
            source,
            INVALID_UNIT_SPECIFIER_ID,
            severity::SEVERITY_ERROR,
            "invalid unit specifier");
    }

    message message_utility::invalid_number_after_radix(const source_reference& source)
    {
        return message(
            source,
            INVALID_NUMBER_AFTER_RADIX_ID,
            severity::SEVERITY_ERROR,
            "invalid number after radix");
    }

    message message_utility::unexpected_end_of_file(const source_reference& source)
    {
        return message(
            source,
            UNEXPECTED_END_OF_FILE_ID,
            severity::SEVERITY_ERROR,
            "unexpected end of file");
    }

    message message_utility::keyword_expected(
        const std::shared_ptr<token>& token, 
        const source_reference& file_source,
        keyword keyword)
    {
        return message(
            token->source(),
            KEYWORD_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "keyword expected");
    }

    message message_utility::symbol_expected(
        const std::shared_ptr<token>& token,
        const source_reference& file_source,
        symbol symbol)
    {
        return message(
            token->source(),
            SYMBOL_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "symbol expected");
    }

    message message_utility::identifier_expected(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            IDENTIFIER_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "identifier expected");
    }

    message message_utility::string_literal_expected(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            STRING_LITERAL_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "string literal expected");
    }

    message message_utility::numeric_literal_expected(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            NUMERIC_LITERAL_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "numeric literal expected");
    }

    message message_utility::integer_literal_expected(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            INTEGER_LITERAL_EXPECTED_ID,
            severity::SEVERITY_ERROR,
            "integer literal expected");
    }

    message message_utility::variable_name_expected_after_type_reference(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            VARIABLE_NAME_EXPECTED_AFTER_TYPE_REFERENCE_ID,
            severity::SEVERITY_ERROR,
            "variable name expected after type reference");
    }

    message message_utility::unexpected_token(
        const std::shared_ptr<token>& token,
        const source_reference& file_source)
    {
        return message(
            token->source(),
            UNEXPECTED_TOKEN_ID,
            severity::SEVERITY_ERROR,
            "unexpected token");
    }
}
